### PR TITLE
boards: Added Teensy-4.x BSP

### PR DESCRIPTION
**Pull request by michallenc**
## Summary

This PR adds board support package for [Teensy 4.0](https://www.pjrc.com/store/teensy40.html) and [Teensy 4.1](https://www.pjrc.com/store/teensy41.html). Both boards have configured I2C, SPI, UART, FlexCAN and ADC, Teensy 4.1 also provides Ethernet connection and micro SD card slot.

## Testing

I2C, SPI, UART, FlexCAN, ADC, Ethernet and SD card tested on Teensy 4.1 or 4.0 (if the functions were available)

### CODE REVIEW

**Comment by patacongo** [comment]
When you add a board, typically you should also update the boards/README.txt, the top-level README.md file,  And rrelevant board documents under Documentation.

**Comment by patacongo** [comment]
Ooops... Sorry.  Didn't mean to close the PR.

**Comment by btashton** [comment]
@michallenc  it's a lot to go through, but I'm excited for this, I have a couple of the 4.0 boards here that have been sitting untouched because I did not want to do the bring up without being able to easily connect a debugger.

You did the hard part!

**Comment by Ouss4** [comment]
@michallenc please run `./tools/refresh.sh --silent teensy-4.x:can-4.1` to fix update the defconfig and fix the builds.

**Comment by michallenc** [comment]
> When you add a board, typically you should also update the boards/README.txt, the top-level README.md file, And rrelevant board documents under Documentation.

Thanks. Apart from README.md a boards/README.txt, the only other change should be done in Documentation/introduction/detailed_support.rst, is it correct?

**Comment by patacongo** [comment]
> 
> 
> > When you add a board, typically you should also update the boards/README.txt, the top-level README.md file, And rrelevant board documents under Documentation.
> 
> @patacongo Thanks. Apart from README.md a boards/README.txt, the only other change should be done in Documentation/introduction/detailed_support.rst, is it correct?

I think so but the documentation has changed so much that I am not certain where things are at anymore.

I personally think that the last part of the top-level README.md and the redundant content of boards/README.txt do not add very much value.  I would personally not be opposed to just removing that kruft and keeping board information in a single place under Documentation.

**Comment by btashton** [comment]
@michallenc This looks good to me.  Thanks for adding the entry for the board to the documentation.

I wont have a 4.1 until next week to test, but I did do basic testing with the 4.0 and nsh
```
NuttShell (NSH) NuttX-10.0.0
nsh> uname -a
NuttX 10.0.0 ad832ed36a Dec  3 2020 21:56:47 arm teensy-4.x
nsh> ls
/:
 dev/
 proc/
nsh> free
             total       used       free    largest
Umem:       781040       6256     774784     774784
nsh> ps
  PID PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK   STACK COMMAND
    0   0 FIFO     Kthread N-- Ready              00000000 002024 Idle Task
    1 100 FIFO     Task    --- Running            00000000 002036 init
nsh> ?
help usage:  help [-v] [<cmd>]

  .         cd        echo      hexdump   mh        rm        time      xd
  [         cp        exec      kill      mount     rmdir     true
  ?         cmp       exit      ls        mv        set       uname
  basename  dirname   false     mb        mw        sleep     umount
  break     dd        free      mkdir     ps        source    unset
  cat       df        help      mkrd      pwd       test      usleep

Builtin Apps:
  sh   nsh
nsh> ls /dev
/dev:
 console
 null
 ttyS0
```

@patacongo Do you have any other concerns?

**Review by btashton — APPROVED** [review]

File: boards/arm/imxrt/teensy-4.x/Kconfig [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	bool "Teensy 4.0"
+
+config TEENSY_41
+	bool "Teensy 4.1"
+
+endchoice # Board configuration
+
+menu "Sensors interfaces"
+
+config BMP280_LPI2C
```
Comment (btashton) [inline boards/arm/imxrt/teensy-4.x/Kconfig:29]: I don't think this should be here.  I don't see it on the board schematics.

**Review by btashton — CHANGES_REQUESTED** [review]

File: boards/arm/imxrt/teensy-4.x/Kconfig (hunk 1 above)
Comment (michallenc) [reply boards/arm/imxrt/teensy-4.x/Kconfig]: @btashton Thanks for your review! Agreed, this was more like a relict from my testing and experimenting.

File: boards/arm/imxrt/teensy-4.x/src/imxrt_flexspi_nor_boot.h [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __BOARDS_ARM_IMXRT_TEENSY_4X_SRC_IMXRT_FLEXSPI_NOR_BOOT_H
+#  define __BOARDS_ARM_IMXRT_TEENSY_4X_SRC_IMXRT_FLEXSPI_NOR_BOOT_H
+
```
Comment (patacongo) [inline boards/arm/imxrt/teensy-4.x/src/imxrt_flexspi_nor_boot.h:23]: This two space indentation is an artifact and an error introduced by tools/indent.sh.  sed -i -e "s/#  /#/g" will fix it.

File: boards/arm/imxrt/teensy-4.x/src/imxrt_flexspi_nor_flash.h [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+#ifndef __BOARDS_ARM_IMXRT_TEENSY_4X_SRC_IMXRT_FLEXSPI_NOR_FLASH_H
+#  define __BOARDS_ARM_IMXRT_TEENSY_4X_SRC_IMXRT_FLEXSPI_NOR_FLASH_H
+
```
Comment (patacongo) [inline boards/arm/imxrt/teensy-4.x/src/imxrt_flexspi_nor_flash.h:22]: Same indentation error introduced by tools/indent.sh.

File: boards/arm/imxrt/teensy-4.x/src/imxrt_flexspi_nor_flash.h [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+/* Definitions for FlexSPI Serial Clock Frequency */
+
+enum flexspi_serial_clkfreq_e
+  {
+    FLEXSPI_SERIAL_CLKFREQ_30MHz = 1,
```
Comment (patacongo) [inline boards/arm/imxrt/teensy-4.x/src/imxrt_flexspi_nor_flash.h:192]: braces should be at column 1.  This is another tools/indent.sh artifact.

File: boards/arm/imxrt/teensy-4.x/src/teensy-4.h [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __BOARDS_ARM_IMXRT_TEENSY_4X_SRC_TEENSY_4_H
+#  define __BOARDS_ARM_IMXRT_TEENSY_4X_SRC_TEENSY_4_H
+
```
Comment (patacongo) [inline boards/arm/imxrt/teensy-4.x/src/teensy-4.h:23]: Same tools/indent.sh problem.

**Review by patacongo — APPROVED** [review]
Looks good!  I did have a few stylistic comments, mostly some weird stuff that got introduced by tools/indent.sh.

**Comment by patacongo** [comment]
> @patacongo Do you have any other concerns?

There is really too much to review here any any real detail.  I did spot a few stylistic things which you may or may not want to addressed.  This is mostly weirdness that was introduced by tools/indent.sh (most described in tools/README).

**Comment by btashton** [comment]
@michallenc if you want to take a pass at the comments Greg made that would be great and then I think we can merge this in.

I did see a couple comments in the board code that looked like they might have been left from another imxrt board like EVKB and OpenSDA. 

I'll have a teensy 4.1 I can use to test some of this other functionality next week after it's in.

Once again thanks for all your hard work on this.

**Comment by patacongo** [comment]
After running indent.sh, I do the following:

      # Run the indent tool on the file.  This requires manual fixups
      # including removing the bogus blank line at the top of the
      # file and and putting the final */ of multi-line comments on a
      # separate line
    
      tools/indent.sh $file
    
      # After indent.sh, fix header files with bad pre-processor indentation

     filename=$(basename $(file))
     extension="${filename##*.}"
    
      if ("X$extension" == "Xh"); then
        sed -i -e "s/#  /#/g" $file
     fi

Do you think this should be added to tools/indent.sh?

File: boards/arm/imxrt/teensy-4.x/src/imxrt_autoleds.c [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/* There are four LED status indicators located on the EVK Board.  The
```
Comment (btashton) [inline boards/arm/imxrt/teensy-4.x/src/imxrt_autoleds.c:21]: There is only the single LED on this board.  I think this comment block needs to be updated.

File: boards/arm/imxrt/teensy-4.x/include/board.h [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+#define GPIO_ENET_RX_ER      (GPIO_ENET_RX_ER_1|IOMUX_ENET_RXERR_DEFAULT) /* GPIO_B1_11 */
+#define GPIO_ENET_TX_CLK     (GPIO_ENET_REF_CLK_2|\
+                              IOMUX_ENET_TX_CLK_DEFAULT)                  /* GPIO_B1_10 */
+#define GPIO_ENET_TX_EN      (GPIO_ENET_TX_EN_1|IOMUX_ENET_EN_DEFAULT)    /* GPIO_B1_09 */
+
+/* PIO Disambiguation *******************************************************/
+
+/* LPUARTs
+ *
+ * Virtual console port provided by OpenSDA on UART1 and
```
Comment (btashton) [inline boards/arm/imxrt/teensy-4.x/include/board.h:232]: This comment block seems to be for a different board

File: boards/arm/imxrt/teensy-4.x/src/imxrt_autoleds.c (hunk 6 above)
Comment (michallenc) [reply boards/arm/imxrt/teensy-4.x/src/imxrt_autoleds.c]: Actually I think there are two of them – user LED and the boot LED (switched on when teensy is in boot mode).

**Comment by michallenc** [comment]
@btashton I pushed the changes @patacongo and you refered to, thanks!

We have our board designed under Teensy 4.1 (https://gitlab.com/pikron/projects/imxrt-devel) for further experiments at faculty. We have there an 1.14 inch Color TFT LCD (st7789 controller) display, which I´d like to get running on Nuttx too. I have i configuration at my nuttx fork, but still having some problems with colours so I will get it here with some further PR when I figure out what´s wrong. We´d also like to get running CDC-ACM as device to have console over USB in future.

**Comment by btashton** [comment]
Looks good.  I have the feather <> teensy adaptor so I was thinking about trying against the featherwing keyboard that I have a PR to get in https://www.tindie.com/products/arturo182/keyboard-featherwing-qwerty-keyboard-26-lcd/  Its nice because it actually uses a lot of peripherals and subsystems.

Feel free to open a draft PR if you are hitting an issue and would like some more eyes on the problem.  There are a few of us active here that use or have imxrt hardware.

**Pull request merged**